### PR TITLE
fix(45901): Error 1064 should apply to @callback definitions in the same way as @param, but doesn't

### LINKED ISSUE #45901: Error 1064 should apply to `@callback` definitions in the same way as `@param`, but doesn't
# Bug Report

I'm using JSDoc to provide type information for JavaScript files. I found a difference in behavior between one-off `@param` definitions and `@callback` definitions where there should not be a difference. Specifically, with `@param`, async functions require a return value of Promise<T>, but with `@callback`, they do not. 

### 🔎 Search Terms

1064, `callback`, async functions, Promise<T>
"The return type of an async function or method must be the global Promise<T> type. Did you mean to write 'Promise<{0}>'?"

### 🕗 Version & Regression Information

I was able to reproduce using Typescript v4.2.3 and v4.4.3. I didn't test other versions.

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about common bugs that aren't bugs, as well doing a text search for `Promise` and `async`.

### ⏯ Playground Link

[Playground link with `@param`](https://www.typescriptlang.org/play?ts=4.4.2&ssl=7&ssc=3&pln=1&pc=1&filetype=js#code/PQKhCgAIUgBAHAhgJ0QW0gbwM4BdkCWAdgOYC+keyUMsyAprgK7JHZZXHk3DgDGAeza5IaAJ4AxIpAC8kRNjFE+lfLIB8WKJB0NmrVcgDc4MkaA) - has error
[Playground link with `@callback`](https://www.typescriptlang.org/play?ts=4.4.2&filetype=js#code/PQKhCgAIUgBBjAhgG2QI0fA1pAsgTwDEA7KGWAB0QCdEBbSAbwGcAXagS2IHMBfSNtTJxqAU1YBXasWZNBXPmWDhwocq3wVRTAiX4hl8APYzWkOkWKQAvJETN8xeAPY2AfEyiRvYydJfUANzgvIFAA) - no error

### 💻 Code

With `@param`:
```js
/**
 * @param {string} str
 * @returns {string}
 */
const myFn = async str => {
    return str;
};
```

With `@callback`:
```js
/**
 * @callback MyFn
 * @param {string} str
 * @returns {string}
 */

/** @type {MyFn} */
const myFn = async str => {
    return str;
};
```

### 🙁 Actual behavior

The version with `@param` produces error 1064, but the version with `@callback` does not.

### 🙂 Expected behavior

Both ways of writing the definition should produce the error.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #45901

### CODE REVIEW

**Comment by sandersn** [comment]
I think the basic approach is fine, and as a targetted fix, this just needs a little cleanup. Couple of points to follow up on, though:

1. I would inline getReturnInfoOfSignatureDeclaration since it's only used once and it allocates (unless the VM optimises it).
2. I thought there was a utility function that checks for a `@type` tag if getEffectiveReturnTypeNode returns undefined. Is that utility appropriate here? It probably doesn't return enough information though -- my memory is that it returns a Type instead of a Node. However, if you can find that utility, you should look at its callers to see if they could improve their error messages if they called getReturnInfoOfSignatureDeclaration instead.

I also need to think about the wording of the new error message. It's a little off but I haven't figured out an improvement.

**Comment by fatcerberus** [comment]
Shouldn't the error message be the same as other instances of this error? i.e.

> The return type of an async function or method must be the global `Promise<T>` type.

I'm not sure why a new error message is even needed.

**Comment by a-tarasyuk** [comment]
@sandersn Thanks for the feedback. There is a utility called `getReturnTypeFromAnnotation` that returns a `Type`. This utility handles cases with JSDoc type tags by utilizing `getReturnTypeOfTypeTag`, which also returns a `Type`. We can attempt to use the type instead of the `TypeNode` to perform all checks. However, we still need the _return_ `TypeNode` for error messages.

> I would inline getReturnInfoOfSignatureDeclaration since it's only used once and it allocates (unless the VM optimises it)

Are you suggesting moving the logic from `getReturnInfoOfSignatureDeclaration` directly into `checkSignatureDeclarationDiagnostics` or relocating `getReturnInfoOfSignatureDeclaration` to the scope of `checkSignatureDeclarationDiagnostics`?

**Comment by sandersn** [comment]
Sorry I dropped this. Now that I look at it again:

1. Move the logic directly into checkSignatureDeclarationDiagnostics.
2. I now disagree with my past self: I don't think it's worthwhile to reuse getReturnTypeFromAnnotation.

**Comment by a-tarasyuk** [comment]
@sandersn Thanks. Do I need to modify the error message for the following?

> The return type of an async function or method must be the global Promise<T> type.

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -199,6 +199,10 @@
         "category": "Error",
         "code": 1064
     },
+    "The return type of an async function is invalid.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json:202]:
```suggestion
    "The return type of an async function or method must be the global Promise<T> type.": {
```

**Review by sandersn — CHANGES_REQUESTED** [review]
re @fatcerberus 's comment: the did-you-mean isn't appropriate since the change needs to be on the callback tag, but the rest of the error still applies.

**Review by sandersn — APPROVED** [review]

**Pull request merged**